argon2/src/instance.rs

```diff
@@ -115,10 +121,54 @@ impl<'a> Instance<'a> {
         Ok(instance)
     }
 
+    /// Create multiple mutable references for the current instance, one for every lane
+    #[cfg(feature = "parallel")]
+    #[allow(unsafe_code)]
+    unsafe fn mut_self_refs(&mut self) -> Vec<&'static mut Instance<'static>> {
+        let lanes = self.lanes;
+        // This transmute can be skipped when a scoped threadpool is used (or when `spawn_unchecked()` gets stabilised)
+        let this = mem::transmute::<_, &mut Instance<'static>>(self);
+        let this: *mut Instance<'static> = this;
+
+        // Dereference the raw pointer multiple times to create multiple mutable references
+        (0..lanes).map(|_| &mut *(this)).collect()
+    }
```

Comment:
Don't want to get too bogged down in this, but I'm curious if there's a safe borrow splitting strategy that can be used here instead

Reply:
It seems like `self.lanes.iter_mut()` could also work here?

Reply:
Hmm, I'm not sure. Looking at the specification, it seems like cross-lane memory access is required

> Mapping J1, J2 to the reference block index. The value l = J2 mod p determines the index of the lane
from which the block will be taken

https://github.com/RustCrypto/password-hashes/blob/e483ce59cf356e9fd1d18b42c61c17b637ff05ed/argon2/src/instance.rs#L245-L260

Off the top of my head I can only think of using either unsafe or a mutex to share the memory between the lanes
The mutex method could work but I'm not sure how much it would impact performance

Reply:
Here's an example of how to do safe mutable borrow splitting:

```rust
fn split_borrows<T>(mut slice: &mut [T]) -> Vec<&mut T> {
    let mut result = vec![];
    
    while let Some((elem, rest)) = slice.split_first_mut() {
        result.push(elem);
        slice = rest;
    }
    
    result
}
```

Reply:
(I was about to reply with the same one-liner to your tweet :) )

I'm sorry but I don't really understand how I should split the memory up to make this work  
Because, as far as I understand, that still unfortunately leaves us with the problem that every lane potentially needs access to every part of the memory 

The memory (even when converted from `&mut [Block]` to an `Vec<&mut Block>`) can't be modified because a vector like this would still only return an `&&mut Block` which can't be used to call a function that needs an `&mut Block` 
(which makes sense, allowing mutations on the underlying data through basically an immutable reference without some kind of lock is a bad idea)

Even though there shouldn't be data races because of how argon2 works, the compiler has no way of knowing that. 
To the compiler it looks we are trying to access memory that another thread could potentially mutate

This could be solved safely by using a mutex but could lead to contention when running the blake2 rounds 
(the `fill_bytes` function could probably be changed to locking the mutex itself instead of requiring a mutable reference to be present beforehand. This should get rid of most of the contention)

Reply:
Okay, sorry, I now realize this is a much thornier problem than I thought it was at first.

I guess I don't have any good suggestions offhand. We can start with this.